udata_hydra/analysis/csv.py

```diff
@@ -0,0 +1,230 @@
+import hashlib
+import json
+import logging
+import os
+import sys
+
+from datetime import datetime
+from typing import Any
+
+import sentry_sdk
+
+from csv_detective.explore_csv import routine as csv_detective_routine
+from progressist import ProgressBar
+from sqlalchemy import MetaData, Table, Column, BigInteger, String, Float, Boolean, Integer
+from sqlalchemy.dialects.postgresql import asyncpg
+from sqlalchemy.schema import CreateTable
+from str2bool import str2bool
+from str2float import str2float
+
+from udata_hydra import context, config
+from udata_hydra.utils.db import (
+    get_check, insert_csv_analysis, compute_insert_query,
+    update_csv_analysis, get_csv_analysis,
+)
+from udata_hydra.utils.file import download_resource
+
+
+log = logging.getLogger("udata-hydra")
+
+# Increase CSV field size limit to maximum possible
+# https://stackoverflow.com/a/15063941
+field_size_limit = sys.maxsize
+while True:
+    try:
+        stdcsv.field_size_limit(field_size_limit)
+        break
+    except OverflowError:
+        field_size_limit = int(field_size_limit / 10)
+
+PYTHON_TYPE_TO_PG = {
+    "string": String,
+    "float": Float,
+    "int": BigInteger,
+    "bool": Boolean,
+}
+
+PYTHON_TYPE_TO_PY = {
+    "string": str,
+    "float": float,
+    "int": int,
+    "bool": bool,
+}
+
+RESERVED_COLS = ("__id", "tableoid", "xmin", "cmin", "xmax", "cmax", "ctid")
+
+
+async def analyse_csv(check_id: int = None, url: str = None, file_path: str = None, debug_insert: bool = False) -> None:
+    """Launch csv analysis from a check or an URL (debug), using previsously downloaded file at file_path if any"""
+    if not config.CSV_ANALYSIS_ENABLED:
+        log.debug("CSV_ANALYSIS_ENABLED turned off, skipping.")
+        return
+
+    assert any(_ is not None for _ in (check_id, url))
+    check = await get_check(check_id) if check_id is not None else {}
+    url = check.get("url") or url
+
+    headers = json.loads(check.get("headers") or "{}")
+    tmp_file = open(file_path, "rb") if file_path else await download_resource(url, headers)
+
+    try:
+        try:
+            inspection_error = None
+            csv_inspection = await perform_csv_inspection(tmp_file.name)
+        except Exception as e:
+            inspection_error = e
+        ca_id = await insert_csv_analysis({
+            "resource_id": check.get("resource_id"),
+            "url": url,
+            "check_id": check_id,
+        })
+        if inspection_error:
+            await handle_parse_exception(inspection_error, "csv_detective", analysis_id=ca_id)
+            return
+        table_name = hashlib.md5(url.encode("utf-8")).hexdigest()
+        await csv_to_db(tmp_file.name, csv_inspection, table_name, debug_insert=debug_insert, analysis_id=ca_id)
+        await update_csv_analysis(ca_id, {
+            "parsing_table": table_name,
+            "parsing_date": datetime.utcnow(),
+        })
+        await csv_to_db_index(table_name, csv_inspection, check)
```

Comment:
I'm wondering if we still need `csv_analysis` then?

Reply:
True! I wondered about it too. I still store the errors there. I could store them in `tables_index`, but it kinda makes sense to have them in the meta tables I think.

Also I'm planning to introduce a `job_id` column to be able to know the status of a long running process (eg analyse csv). This could be a good place to store that.

I guess this should be decided/refined when we build APIs, both on the data and on the metadata. Maybe keep it there for now? 😇

Reply:
I'm okay with that :)